pkg/agent/controller/networkpolicy/packetin.go

```diff
@@ -187,3 +223,98 @@ func getPacketInfo(pktIn *ofctrl.PacketIn, ob *logInfo) error {
 	}
 	return nil
 }
+
+// rejectRequest sends reject response to the requesting client, based on the
+// packet-in message.
+func (c *Controller) rejectRequest(pktIn *ofctrl.PacketIn) error {
+
+	// Get ethernet data.
+	srcMAC := pktIn.Data.HWDst
+	dstMAC := pktIn.Data.HWSrc
+
+	var (
+		srcIP    string
+		dstIP    string
+		prot     uint8
+		ipHdrLen uint16
+	)
+	switch pktIn.Data.Ethertype {
+	case protocol.IPv4_MSG:
+		// Get IP data.
+		ipv4In := pktIn.Data.Data.(*protocol.IPv4)
+		srcIP = ipv4In.NWDst.String()
+		dstIP = ipv4In.NWSrc.String()
+		prot = ipv4In.Protocol
+		ipHdrLen = IPv4HdrLen
+	case protocol.IPv6_MSG:
+		// Get IP data.
+		ipv6In := pktIn.Data.Data.(*protocol.IPv6)
+		srcIP = ipv6In.NWDst.String()
+		dstIP = ipv6In.NWSrc.String()
+		prot = ipv6In.NextHeader
+		ipHdrLen = IPv6HdrLen
+	}
+
+	pktOutBuilder, err := c.getBasePacketOutBuilder(srcMAC.String(), dstMAC.String(), srcIP, dstIP)
+	if err != nil {
+		return err
+	}
+
+	switch prot {
+	case protocol.Type_TCP:
+		// Get TCP data.
+		TCPSrcPort, TCPDstPort, TCPSeqNum, TCPAckNum, err := getTCPHeaderData(pktIn.Data.Data)
+		if err != nil {
+			return err
+		}
+		return c.ofClient.SendTCPReject(pktOutBuilder, TCPSrcPort, TCPDstPort, TCPSeqNum, TCPAckNum, false)
+	default: // Use ICMP host administratively prohibited for ICMP, UDP, SCTP reject.
+		ipHdr, _ := pktIn.Data.Data.MarshalBinary()
+		ICMPData := make([]byte, int(ICMPUnusedHdrLen+ipHdrLen+8))
+		// Put ICMP unused header in Data prop and set it to zero.
+		binary.BigEndian.PutUint32(ICMPData[:ICMPUnusedHdrLen], 0)
+		copy(ICMPData[ICMPUnusedHdrLen:], ipHdr[:ipHdrLen+8])
+		return c.ofClient.SendICMPReject(pktOutBuilder, ICMPData, false)
+	}
+}
+
+// getTCPHeaderData gets TCP header data used in the reject packet.
+func getTCPHeaderData(ipPkt util.Message) (TCPSrcPort uint16, TCPDstPort uint16, TCPSeqNum uint32, TCPAckNum uint32, err error) {
+	var tcpBytes []byte
+
+	// Transfer Buffer to TCP
+	switch ipPkt.(type) {
+	case *protocol.IPv4:
+		tcpBytes, err = ipPkt.(*protocol.IPv4).Data.(*util.Buffer).MarshalBinary()
+	case *protocol.IPv6:
+		tcpBytes, err = ipPkt.(*protocol.IPv6).Data.(*util.Buffer).MarshalBinary()
+	}
+	if err != nil {
+		return 0, 0, 0, 0, err
+	}
+	tcpIn := new(protocol.TCP)
+	err = tcpIn.UnmarshalBinary(tcpBytes)
+	if err != nil {
+		return 0, 0, 0, 0, err
+	}
+
+	return tcpIn.PortDst, tcpIn.PortSrc, 0, tcpIn.SeqNum + 1, nil
+}
+
+// getBasePacketOutBuilder gets a base IP packetOutBuilder with src&dst MAC, IP and OFPort settled.
```

Comment:
what does "settled" mean here?

Reply:
It means these props of this IP PacketOutBuilder have been assigned.

Reply:
I think you mean "set" then. "settled" has different meanings.